### DISCUSSION TITLE
How to display ancestors, not just first section?

### DISCUSSION THREAD

**Discussion by TetraTheta** [hbstack/.github#63] [Q&A] [answered]
Sorry for the vague title. I couldn't figure out better title for this problem.

***

By default, HB Cards Theme shows Topmost Section next to Estimated Read Time. I want to display the page's ancestors like 'Section 1 / Section 1-2 / Section 1-2-3', so I modified [`first-section.html`](https://github.com/hbstack/blog/blob/main/layouts/partials/hb/modules/blog/post/meta/first-section.html) like this:
```html
{{- /* Don't use .FirstSection (everything become 'Blog') */ -}}
{{- with .CurrentSection -}}
<span class="hb-blog-post-section">
  <div class="badge bg-secondary text-decoration-none fw-normal">
    {{- range .Ancestors.Reverse -}}
      {{- if or .IsHome (eq .Title "Blog") -}}
        {{- continue -}}
      {{- end -}}
    <a class="bg-secondary text-decoration-none fw-normal" style="color: var(--hb-badge-color);" href="{{ .RelPermalink }}">{{ .Title }}</a>
    <div class="bg-secondary text-decoration-none fw-normal" style="display: inline-block; color: var(--hb-badge-color); opacity: 0.45;">&nbsp;/&nbsp;</div>
    {{- end -}}
    <a class="bg-secondary text-decoration-none fw-normal" style="color: var(--hb-badge-color);" href="{{ .RelPermalink }}">{{ .Title }}</a>
  </div>
</span>
{{- end -}}
```
I didn't want to show 'Blog' in sections, I added corresponding code too.

Here is a breif summary of this modified code.
> The actual ancestors is like 'Blog / Section 1 / Section 1-2 / Section 1-2-3'. But like I said, I removed 'Blog' from the ancestors, so it will be like 'Section 1 / Section 1-2 / Section 1-2-3.
> If you click 'Section 1', you'll go to Section 1(`/blog/section-1/`). If you click 'Section 1-2', you'll go to Section 1-2(`/blog/section-1/section-1-2/`), and so on.

But recent change broke my code. I tried to update my code to fit latest update, but I couldn't.

How can I fix my code to display things as it used to do?

***

PS. Should I create my own partial `ancestors.html` and `ancestors-reverse.html` for removing 'Blog' entry from `.Ancestors`? I'm doing same thing on Breadcrumb too, so I wonder if it aid performance or not.

**Comment by razonyang** [comment]
I'm afraid not possible, since it was cached by the firstSection.

https://github.com/hbstack/blog/blob/2b9c27f3353f66cb7c7a6e7852e4a066592769c3/layouts/partials/hb/modules/blog/post/card.html#L40

I can treat it a feature request instead, are you want to display the closest parent section instead?

**Reply by TetraTheta** [reply]
I want to display all ancestors except 'Blog' (not 'Docs', if present) as 'Section-1 / Section-1-2'. I'm already displaying closest parent section as categories, so displaying closest parent section as 'first-section' would look like duplicated.

I think I must override `card.html` too. Since I revert back to `partial` from `partialCached`, this will reduce performance gain from recent commits, though...

**Reply by razonyang** [reply]
> I think I must override card.html too. Since I revert back to partial from partialCached

It's a workaround, but it would become a maintain issue, you'll have to check the changes and suit your needs.

**Reply by TetraTheta** [reply]
It would be very nice to have 'my needs' to be implemented as official feature, but I'm afraid that would be too niche use cases. That's why I decided to override layout files myself, even though it would lead this kind of problem - my workaround got broken after update -.

For example, the way I want to display 'first-section' will decrease performance because Hugo now have to dig through all of the pages' ancestors and filter them. But because I really wanted it anyway, I choose to override it.

You can consider my words as Feature Request, but I'm really afraid that my overrides would be considered as 'too niche use cases' and get rejected.

**Comment by razonyang** [comment]
> Should I create my own partial ancestors.html and ancestors-reverse.html for removing 'Blog' entry from .Ancestors

Is there any reason to remove `Blog`? I think delete `content/blog/_index.md` achieves same goal.

> I'm doing same thing on Breadcrumb too, so I wonder if it aid performance or not.

Yes, the newer changes cache partials which will produce same output as far as possible.

**Reply by TetraTheta** [reply]
I don't know well about partial cache, so I still wonder how will it affect performance. I haven't used it before...

As far as I know, 'filtered ancestors' will be used at 'Pages(for Blog, Docs, Series, Categories etc. pages)' and 'Article(article itself, for breadcrumb)'. So using `partialCached` in both `first-section.html`(for 'Pages') and `breadcrumb/index.html`(for Breadcrumb) will help performance, right?

> Is there any reason to remove `Blog`? I think delete `content/blog/_index.md` achieves same goal.

I tried to rename `content/b/_index.md`(this is actual path) to `@@@@@@_index.md` to make it not recognized as `_index.md`. `first-section` becomes `Bs`. This looks like 'Blog' becomes 'Bs'.

I'm not 100% sure, but I think I need that `_index.md`...

**Reply by razonyang** [reply]
> so I still wonder how will it affect performance

It just cache and reuse the output for pages, for example, section pages (list) `blog/foo/bar/` and `blog/foo/bar/page/2`, has same breadcrumb (`home / blog / foo / bar`), caching the output with current section as the cache key, then reuse it for subsequence pages, so that theme won't have to execute template multiple time for pages has same breadcrumb.

If you've custom breadcrumb, you can tweak the [layouts/partials/hugopress/modules/hb-blog-breadcrumb/hooks/hb-blog-main-begin.html](https://github.com/hbstack/blog/blob/main/modules/breadcrumb/layouts/partials/hugopress/modules/hb-blog-breadcrumb/hooks/hb-blog-main-begin.html), changes on here https://github.com/hbstack/blog/commit/590bbd030912171495d6d71a929a3bf2bf0e272b.

**Reply by razonyang** [reply]
But I think this change may not affect your custom breadcrumbs, it just caching for **non single/detailed pages**

**Reply by TetraTheta** [reply]
If it only affects 'non-single/detailed pages'('Pages(for Blog, Docs, Series, Categories etc. pages)' that I said), then it is not what I want to modify instead. Breadcrumbs shows up in both 'non-single/detailed pages' and 'single/detailed pages'. I want to remove 'Blog' from both area.

But I'll try it after I create `ancestors.html` partial code myself to see if it works for both area or not.

**Reply by razonyang** [reply]
You can revert the [layouts/partials/hugopress/modules/hb-blog-breadcrumb/hooks/hb-blog-main-begin.html](https://github.com/hbstack/blog/blob/main/modules/breadcrumb/layouts/partials/hugopress/modules/hb-blog-breadcrumb/hooks/hb-blog-main-begin.html)'s [changes](https://github.com/hbstack/blog/commit/590bbd030912171495d6d71a929a3bf2bf0e272b) by creating it under your project if it doesnt meet your needs.

**Comment by razonyang** [comment, marked as answer]
With [github.com/hbstack/blog@v0.36.0](https://github.com/hbstack/blog/releases/tag/v0.36.0), you can disable the first section by configuring cascade.

```yaml
// hugo.yaml
cascade:
  - _target:
      path: /**
    params:
      meta:
        first_section: false
```

And then create `hb-blog-post-meta` hook, and save it as `layouts/partials/hugopress/modules/hb-custom/hooks/hb-blog-post-meta.html`, please do not forgot to configure it in params.yaml to tell hooks system to load it.

```yaml
// params.yaml
// ...
hugopress:
  modules:
    hb-custom:
      // ...
      hooks:
        hb-blog-post-meta:
          cacheable: false
```

With this, you don't have to care about upstream changes, you can append any meta you want.

**Reply by TetraTheta** [reply]
Thank you! I'll try this right now. I'll come back after I check everything works as intended.

**Reply by razonyang** [reply]
Please note that, the hook context is different, to access page then use `.Page`.

**And this way may show the meta on both list and single page.**

**Reply by TetraTheta** [reply]
I think I can determine if the `.Page` object is actual page by [`.Page.IsPage`](https://gohugo.io/methods/page/ispage/).

**Reply by razonyang** [reply]
```go
{{ if page.IsSection }}
  Meta only on list
  {{ .Page.CurrentSection.Title }}
{{ end }}
```

Please upgrade to v0.36.1 (context fix), and use `page` to determind parent page's kind, `.Page` is current render page of particular card, `.Page.IsPage` is always `true`.

**Reply by TetraTheta** [reply]
I couldn't figure out how to migrate to a new layout. I set my blog source as public temporarily and pushed [my attempt](https://github.com/TetraTheta/TetraLogSource/commit/b3a9689e0a82a17e869af834e1754b10cfb8f39f) to it. To test this, please do shallow clone with `git clone ~~ --depth 1` to minimize time to download repo.

I use [the current production site](https://tetralog.haipa.xyz/) as a _base_. If something is different than the _base_, it is a bug that I have to resolve.

I found a few problems with my current code:
- Page Meta isn't consistent. Depending on where you are looking at the Page Meta, the content of Page Meta changes. It may shows up or not.
- Breadcrumbs still shows 'Blog'('블로그') entry even though I tried to remove it.
- Header Navigation Menu isn't highlighted when in sub-menu.  
  In the _base_, Header Navigation Menu isn't highlighted in the article page. I think this is a bug that I missed in previous updates.
- (And other problems that I might miss)

There might be other issues that I haven't found yet.

**Reply by razonyang** [reply]
> Page Meta isn't consistent. Depending on where you are looking at the Page Meta, the content of Page Meta changes. It may shows up or not.

The `page` global function seems not consistent for some reason (maybe misused in somewhere), use CSS to hide the page meta from single page instead.

> Breadcrumbs still shows 'Blog'('블로그') entry even though I tried to remove it.

Because you removed your custom `layouts/partials/hb/modules/breadcrumb/index.html`, bring it back and tweak it to suit your needs.

> Header Navigation Menu isn't highlighted when in sub-menu.
In the base, Header Navigation Menu isn't highlighted in the article page. I think this is a bug that I missed in previous updates.

The `is-crrent` function is abandoned, the menus state are detected by JavaScript now, to save build time.

Image that if you have 20 menus (including submenus), and 100k pages, then it will calculate the active state 20 * 100k (2 million) times, it will save at least 2*`n` million operations (`n` represents the template directives, such as assignments, comparisions, string operations and so on).

I created a patch for this, see https://github.com/hbstack/header/pull/623, it will find detailed page's parent menu by URLs from 1) second-level menus (high priority) and 2) top-level menus, once parent menu was found, then stop, that is, there will be up to only one parent menu to be active.

If it doesn't meet your needs, please create and tweak the JS, see https://github.com/hbstack/header/pull/623/files.

> (And other problems that I might miss)

Please open new issue per problem, to track issues easilly.

I'm pushing changes to [private repo](https://github.com/razonyang/TetraLogSource) (I've invited you, I don't have permission to create PR to your repo), you can check out changes (**git clone and git checkout -b patch-1**) about how to solve it.

**Reply by TetraTheta** [reply]
> > Page Meta isn't consistent. Depending on where you are looking at the Page Meta, the content of Page Meta changes. It may shows up or not.
> 
> The page global function seems not consistent for some reason (maybe misused in somewhere), use CSS to hide the page meta from single page instead.

Sorry for the misleading information. I was saying that sometimes Page Meta is 'Category 1 / Category 1-2', but sometimes it is 'Category 1-2' or nothing. When I see Page Meta from site root, Page Meta is blank. But when I manually navigate to those subdirectory like `localhost/b/category-1/category-1-2', I can see Page Meta is showing, but not 100% accurate.

> I created a patch for this, see https://github.com/hbstack/header/pull/623, but I think you'll need to tweak it to suit your particular needs... See the file change, you maybe got inspired hwo to implement.

Should I override(copy to local and edit) this `assets/hb/modules/header/js/index.ts` file to change behavior to fit my needs, like I currently do to HB Card Theme layout files?

I think partially modifying it isn't possible...

> I don't have permission to create PR to your repo

Oops, I'm sorry for your inconvenience. That should be remnant of setting that the repo was private. I checked setting of the repo, but I couldn't find out how to enable PR publically.

Because GitHub's limited Git LFS bandwidth quota(1 GiB per month) is very small to me, I'm planning to migrate to self-hosted Forgejo instance to have _(almost) limitless_ bandwidth quota in very near future.  
No GitLab because they are planning to limit storage(repo + LFS) size to 5GiB per 'namespace'. No BitBucket because storage limit(1 GiB) is far less than GitLab's.

I think this will make you uncomfortable for inspecting files of my blog source. I'm sorry about this inconvenience in advance.

**Reply by razonyang** [reply]
> Should I override(copy to local and edit) this assets/hb/modules/header/js/index.ts file to change behavior to fit my needs, like I currently do to HB Card Theme layout files?

Try [github.com/hbstack/header@v0.14.6](https://github.com/hbstack/header/releases/tag/v0.14.6) and see if it meet your needs, I modified that patch a lot.

If not, you may have to [copy the js](https://github.com/hbstack/header/blob/main/assets/hb/modules/header/js/index.ts) and save it as `assets/hb/modules/header/js/index.ts`.

**Reply by razonyang** [reply]
> Sorry for the misleading information. I was saying that sometimes Page Meta is 'Category 1 / Category 1-2', but sometimes it is 'Category 1-2' or nothing. When I see Page Meta from site root, Page Meta is blank. But when I manually navigate to those subdirectory like `localhost/b/category-1/category-1-2', I can see Page Meta is showing, but not 100% accurate.

Perhaps caching issue, restart Hugo server when you've changed templates.
And the `page` global function seems not consistent for some  reason, you can just hide the your page section from detail pages by CSS, see https://github.com/razonyang/TetraLogSource/commit/e40db832768a76cacc190d80046a7d199a8c57e5

**Reply by TetraTheta** [reply]
> Try [github.com/hbstack/header@v0.14.6](https://github.com/hbstack/header/releases/tag/v0.14.6) and see if it meet your needs, I modified that patch a lot.
If not, you may have to [copy the js](https://github.com/hbstack/header/blob/main/assets/hb/modules/header/js/index.ts) and save it as assets/hb/modules/header/js/index.ts.

I think I have to manually override the file. In my site, menu layout isn't same with actual URL structure.
```plaintext
// Menu structure
A
B
C
D
// Actual file structure
A
B
├ C
└ D
```
So I excluded C/D in submenu of B, and via `is-current.html`, I manually unmark B as active if page URL contains URL for C/D (and B).

Thank you for the clarification!

PS. How did you set `lib` compiler option of Typescript to? I get warning about `Array.from()` that there is no `from` attribute from VSCode.

***

This is what I meant for 'Inconsistent Page Meta'.

![1](https://github.com/hbstack/.github/assets/37061801/b72f99ca-8f35-4528-8573-e1a449903708)  
**↑ When on site root**

![2](https://github.com/hbstack/.github/assets/37061801/7763bbea-613e-441f-9f81-772be55113dd)
**↑ When on Menu B**

![3](https://github.com/hbstack/.github/assets/37061801/8f389f2c-c755-4fca-99b4-355d6892a572)
**↑ When on Menu C (actual URL is under Menu B)**

![4](https://github.com/hbstack/.github/assets/37061801/2c8637fb-05e7-424b-a324-6f9f32bebc80)
**↑ When on submenu of Menu C**

All of these points to same article, but the page meta isn't same like before.

I've tried `.`, `.Page`, `page` but I don't get what should I use in this case.

**Reply by razonyang** [reply]
> This is what I meant for 'Inconsistent Page Meta'.

Just do not use `page`, use `.Page` for current page of card.

![image](https://github.com/hbstack/.github/assets/17720932/88249cd7-b208-4075-816c-fe85b79ba2ac)

![Peek 2024-06-16 20-27](https://github.com/hbstack/.github/assets/17720932/c82793d7-fcc6-4ae5-83c3-656136d0d644)

**Reply by razonyang** [reply]
As to the menu states, you'll need to either a) change structure of your site or b) change the JS file to suit your particular needs...

> How did you set lib compiler option of Typescript to? I get warning about Array.from() that there is no from attribute from VSCode.

Google it or just ignore it, since it just a warning not an error.

**Reply by TetraTheta** [reply]
Thanks to your advise, I've managed to figure out Page Meta with this code:
```html
{{/* hb-blog-post-meta.html: Custom page meta */}}
{{- $ancestors := partial "hb-custom/functions/ancestors" .Page -}}
<span class="hb-blog-post-section">
  <div class="badge bg-secondary text-decoration-none fw-normal">
  {{- range $i, $a := $ancestors -}}
    <a class="bg-secondary text-decoration-none fw-normal" style="color: var(--hb-badge-color);" href="{{ $a.RelPermalink }}">{{ $a.Title }}</a>
    {{- if ne $i (sub ($ancestors | len) 1) -}}
    <div class="bg-secondary text-decoration-none fw-normal" style="display: inline-block; color: var(--hb-badge-color); opacity: 0.45;">&nbsp;/&nbsp;</div>
    {{- end -}}
  {{- end -}}
  </div>
</span>
```
> Google it or just ignore it, since it just a warning not an error.

I decided to use `//@ts-ignore` to suppress the warning. Thank you for the information!

***

I couldn't figure out how to active menu items as I like. Contrary to [the image above](https://github.com/orgs/hbstack/discussions/63#discussioncomment-9790147), the menu isn't get highlighted as I wanted.
```typescript
(() => {
  document.addEventListener('DOMContentLoaded', () => {
    const header = document.querySelector('.hb-header') as HTMLElement
    const nav = header.querySelector('.hb-header-menus') as HTMLElement // only include menu items
...
    // filter games
    const checkPath = (menu: Element) => {
      const filter = ['genshin-impact', 'wuthering-waves']
      const href = window.location.href // for later use
      const pathname = window.location.pathname // for later use
      const menuHref = menu.getAttribute('href') ?? ''
      // I only interested in 'game' menu
      if (pathname.indexOf('game') < 0) return true
      let isDuplicated = false
      filter.forEach(f => {
        if ((pathname.indexOf(f) >= 0) && (menuHref.indexOf(f) < 0)) isDuplicated = true
      })
      return !isDuplicated
    }
...
    const menus = nav.querySelectorAll(`a[href="${window.location.href}"], a[href="${window.location.pathname}"]`)
    if (menus.length > 0) {
      menus.forEach(menu => {
        // activeMenu(menu)
        if (checkPath(menu)) activeMenu(menu) // all three occurance of 'activeMenu(menu)' is replaced to 'if (checkPath(menu)) activeMenu(menu)'
      })
    } else {
...
```
![1](https://github.com/hbstack/.github/assets/37061801/d25a9b0d-973e-4b18-a4bb-e197f3bb984c)

**↑ Without `index.ts` override**

![2](https://github.com/hbstack/.github/assets/37061801/e0e981c5-78b1-43e5-9f05-f41605a437b2)

**↑ With my `index.ts` override**

![3](https://github.com/hbstack/.github/assets/37061801/31a1ded0-66f9-4fb1-b053-3202e820e695)

**↑ What I expects it to be**

Something is wrong about my logic, but I can't figure it out. This should prevent activating '게임' menu if `window.location.pathname` contains either `genshin-impact` or `wuthering-waves`, while letting '명조', '원신' menu get activated.

I sincerely apologize for my inexperience and nagging you.

**Reply by razonyang** [reply]
Have you pushed your code into remote repo? So that I can pull and check.

> I sincerely apologize for my inexperience and nagging you.

No problem.

**Reply by TetraTheta** [reply]
I just pushed a commit that contains every modification I've done so far.

If you have any problem for creating PR, please leave a comment in Discussion of my repository or here. I think I've done everything that I can do about it, but I'm not 100% sure about that.

**Reply by razonyang** [reply]
I dont see any problem on both of `↑ Without index.ts override` and `With index.ts override`... Have you try removing the caches and restarting Hugo server?

The following video are captured with using built-in header module (without index.ts override).

![Peek 2024-06-17 18-10](https://github.com/hbstack/.github/assets/17720932/3dcafee5-a483-4401-bf22-3f7b6fea9993)

There will be only one top-level be actived for detail pages, if not, there may be a cache issue.

**Reply by TetraTheta** [reply]
I removed my override by renaming it to `.ignore.index.ts.txt` which should be enough. I also purged browser cache and any other purgable things too.

![a](https://github.com/hbstack/.github/assets/37061801/a9a02aac-ecc2-4e70-bfc9-0d3d78b01e08)

I really don't understand why this inconsistency happens. I checked update hugo module updates by running `hugo mod get -u ./...` but there is no update available.

**Reply by razonyang** [reply]
So my video shown is what you want? If so, I'd recommend starting Hugo server on other port (`hugo server -p 12345`), and using other browser to test.

What I did is just rename the custom one, to use built-in JS.

![image](https://github.com/hbstack/.github/assets/17720932/1d66ee9f-4ee9-4b1a-a218-a38037153b3f)

Btw, the latest commit I used is `0862460e624810a59406ad83c455566de8c35cc8` which message is `[WIP] Fix layouts`

**Reply by TetraTheta** [reply]
Yes, what you've see locally is what I want.

I removed `-p 80` part from the command I've been using, and it now magically works as intended. Then, I added it back. Now it breaks as it previously did.

But I wonder will it cause problem with build result (and actual site). Something works on `localhost:1313` but not on `localhost:80`(and `localhost`) doesn't look natural to me. 

Do you also experience same problem when running on port 80?

**Reply by razonyang** [reply]
> Do you also experience same problem when running on port 80?

One possible clue is that the previous server process wasn't exited, still running... So other server run on same port will be forced to use other ports, then you're testing anything on the old server.

**Reply by TetraTheta** [reply]
From my (little) experiences, I don't think that is this case.  
When port collision happens, Hugo process opened later will immediately be closed because it couldn't assign port it wants. I've experienced new Hugo window get immediately closed because old Hugo instance was already running few times.

I cleaned 'StandbyList' of Windows by running 'EmptyStandbyList', cleaned browser cache, made sure that there is no Hugo process is running by executing `taskkill /f /im hugo.exe`, then run Hugo server with port 80. Wrong header menu is highlighted again.

I will try to nuke all Hugo cache by deleting corresponding directory.

**Reply by razonyang** [reply]
Have no idea, I rarely develop on Windows, I'd recommend using other port when customizing templates to avoid encountering unexpected cases.

**Reply by TetraTheta** [reply]
Can you check if there is any check for 'insecure connection' among npm library you use/depend?

I tested random ports like `55`(less than `1000`), `443`(HTTPS), `65500`(very high port number), `80`(HTTP). Only `80` had problem. In general, connection to port `80` is HTTP, which is insecure. I think there might be a condition that checks if connection is secure or not by checking port `80`, rather than comparing URL scheme etc.

I'm not fan of using other ports than `80` when local test, but I'll use other port until I can figure out why this happens... Thank you for the massive help!

**Reply by razonyang** [reply]
> Can you check if there is any check for 'insecure connection' among npm library you use/depend?

They are irrelevant, those packages are used to generate files (JS, CSS), nothing related to insecure connection, Google it may help, I didn't encouter related issues myself, couldn't help.

**Reply by TetraTheta** [reply]
Maybe I should change browser to test if this is only for current browser or not...

Thanks for the help! ❤
